### PR TITLE
Adding support for bosonic path integral molecular dynamics simulations

### PR DESCRIPTION
**Pull request by yotamfe**
**Summary**

Introducing two new fixes: **pimdb/langevin** and **pimdb/nvt**, which extend pimd/langevin and pimd/nvt (respectively). The new fixes perform path integral molecular dynamics simulations for bosons, taking bosonic exchange into account, using a quadratic scaling algorithm (described in [doi.org/10.1063/5.0173749](https://doi.org/10.1063/5.0173749)).

Thanks for considering our contribution!

**Related Issue(s)**

None.

**Author(s)**

Ofir Blumer
Tel Aviv University, Tel Aviv, Israel
@OfirBlumer

Jacob Higer
Tel Aviv University, Tel Aviv, Israel
@higj

Yotam Feldman
Tel Aviv University, Tel Aviv, Israel
@yotamfe 
yotam.feldman@gmail.com

Barak Hirshberg
Tel Aviv University, Tel Aviv, Israel
@BarakHirshberg 
barak.hirshberg@gmail.com

We gratefully acknowledge consultation from Yifan Li @Yi-FanLi.

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

The new pimdb/* fixes, and minor changes to the existing pimd/*, do not break backward compatibility.

**Implementation Notes**

Bosonic and ordinary, distinguishable path integral molecular dynamics were tested extensively on a system of trapped particles, and on free particles with periodic boundary conditions. Additional tests for correctness of input validation etc. are listed in a [Google Sheet](https://docs.google.com/spreadsheets/d/1Cmf-mZo6p0kHEJ9r52GmIZdzCuVI39M-1cpjCRLBr8Q/edit?usp=sharing). 

**Post Submission Checklist**

- [ ] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

List of tests: [Google Sheet](https://docs.google.com/spreadsheets/d/1Cmf-mZo6p0kHEJ9r52GmIZdzCuVI39M-1cpjCRLBr8Q/edit?usp=sharing).

### CODE REVIEW

**Comment by akohlmey** [comment]
@yotamfe Thanks for your submission. There are a bunch of "technical" changes required to better comply with the LAMMPS programming style conventions and make the code and documentation translate correctly on a variety of platforms. Since you seem to have an "institutional" GitHub account, I cannot push directly into your submitted branch and thus have submitted a pull request into your LAMMPS fork at: https://github.com/yotamfe/lammps/pull/2

**Comment by yotamfe** [comment]
Thanks so much @akohlmey! I thought edits by maintainers would be OK, but it's probably still problematic if my own fork is from an organization, I'll know for next time. I've integrated your PR. Thanks for all the corrections and improvements!

File: src/REPLICA/bosonic_exchange.h [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  void spring_force_first_bead(double **f) const;
+  void spring_force_interior_bead(double **f) const;
+  void Evaluate_VBn();
+  void Evaluate_V_backwards();
+
+  const int nbosons;
+  const int np;
+  const int bead_num;
+  const bool apply_minimum_image;
+  // In the "reduced-beta convention" [e.g. in J. Chem. Phys. 133, 124104 (2010); also J. Chem. Phys. 74, 4078-4095 (1981)],
```
Comment (akohlmey) [inline src/REPLICA/bosonic_exchange.h]: Please move this comment to a suitable location in the corresponding .cpp file.
We prefer to have only short comments describing in the headers unless they contain doxygen documentation.

**Comment by akohlmey** [comment]
It would be nice to have some minimal/fast examples using these fixes, similar to what is already available in `examples/PACKAGES/pimd`. Those would demonstrate the usage, but could also be used for regression testing or to check for memory issues with valgrind.

File: src/REPLICA/bosonic_exchange.h (hunk 1 above)
Comment (yotamfe) [reply src/REPLICA/bosonic_exchange.h]: Thanks, moved in [8780cd8](https://github.com/lammps/lammps/pull/4479/commits/8780cd89a569eebc2a872ff7f8904d12c6bf60b0)

**Comment by yotamfe** [comment]
> It would be nice to have some minimal/fast examples using these fixes, similar to what is already available in `examples/PACKAGES/pimd`. Those would demonstrate the usage, but could also be used for regression testing or to check for memory issues with valgrind.

That's a great idea. How fast should the examples be? Asking because if we want to have a significant deviation from ordinary PIMD (for distinguishable particles) it might take a bit longer, we'll try to see how fast we can make this.

**Comment by akohlmey** [comment]
> Asking because if we want to have a significant deviation from ordinary PIMD (for distinguishable particles) it might take a bit longer, we'll try to see how fast we can make this.

It doesn't have to be scientifically meaningful. Just show some typical command options and execute the corresponding functions in the code so we can check for memory leaks and incorrect accesses. If is often a good idea to provide a README file that explains some of the choices from the practical point for a given scenario (and thus complement the manual which is more like a command reference) and can also explain how to modify the input for a meaningful run.

Typical examples run for 100 to 1000 steps, rarely much longer. Depending on the choice of potential, the results may diverge on different platforms with different compilers due to floating-point math rounding limitations and different compilers leading to different ordering of operations and different degrees of vectorization.

**Comment by Yi-FanLi** [comment]
Hi Yotam, it looks good to me at a first glance. I will take a more careful look later soon. I just want to confirm one thing: in our email thread last time, we discussed about applying pimd/langevin to a specific group. This part is still to be done by me and this PR does not involve that point, am I correct? @OfirBlumer 

Thanks!

**Comment by yotamfe** [comment]
Great, we'll get it done this week. Thanks!

**Comment by yotamfe** [comment]
Hi Yifan, thanks for looking into it! Yes, this PR doesn't involve this, we
were worried at the point about not breaking this feature if it exists.
It's a good feature in my opinion, but not part of this PR.

**Comment by OfirBlumer** [comment]
Hi all,
Thanks for all the help!
I've just added two example input files for 100 timesteps-long simulations of bosons in a harmonic trap.
A README file explains how they can be extended to longer simulations providing physical quantities that can be compared with an analytic solution.

**Comment by sjplimp** [comment]
@yotamfe To echo @akohlmey, thank you for this contribution with a new capability for LAMMPS.   It all looks good to me.  I have the following suggestions and question (3):

(1) A more conventional LAMMPS name for the new fixes would be fix pimd/boson/nvt (or possibly fix pimd/nvt/boson), ditto for the langevin variants.  With corresponding names for the src files.  Likewise examples/PACKAGES/pimd_boson for the example scripts.  This is to make it more clear for users what the relationship is between pimd and pimd/boson and that the latter derives from the former.

(2) For the 2 doc pages: fix pimd and and fix pimd/boson.  I suggest the original fix pimd page list all 4 commands (2 old, 2 new) at the top.  With your 2 new commands simply being a link to your new doc page.  Similar to how the dump doc page has links at the top to separated doc pages for image, vtk, etc.

Then in the 1st or 2nd Description paragraph of the old doc page (PIMD), say that this page covers standard pimd (briefly explain) and this other page (give link) covers bosonic pimd (briefly explain).  Do the same in the 1st/2nd Description paragraph of your new page. The idea is to give users enough info to understand which commands and doc page they should be looking at.  Likewise to advertise to them the existence of the new variants.  Both doc pages should point to the respective examples/PACKAGES dirs for concrete examples.

No other changes needed to the 2 doc pages. 

(3) For the old fix pimd doc page, you made some edits.  It's not clear to me if these were simply cosmetic or clarification changes, or if you are also changing some feature or option in the fix pimd source files?

**Comment by OfirBlumer** [comment]
@sjplimp Thank you for your kind review of our code.
We accepted your suggestions in (1) and (2) and updated the code, example files, and documentation accordingly.
As for the question in (3), the changes in the documentation are indeed for clarification, and no feature or option in the fix pimd source files was changed.

File: doc/src/fix.rst [hunk 2]
```diff
@@ -341,6 +341,8 @@ accelerated styles exist.
 * :doc:`phonon <fix_phonon>` - calculate dynamical matrix from MD simulations
 * :doc:`pimd/langevin <fix_pimd>` - Feynman path-integral molecular dynamics with stochastic thermostat
 * :doc:`pimd/nvt <fix_pimd>` - Feynman path-integral molecular dynamics with Nose-Hoover thermostat
+* :doc:`pimd/langevin <fix_pimd_bosonic>` - Bosonic Feynman path-integral molecular dynamics for with stochastic thermostat
```
Comment (sjplimp) [inline doc/src/fix.rst]: Should be pimd/langevin/bosonic and similar for nvt/bosonic

File: doc/src/fix_pimd_bosonic.rst [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+   fix 1 all pimd/nvt/bosonic method pimd fmass 1.0 sp 1.0 temp 2.0 nhc 4
+   fix 1 all pimd/langevin/bosonic integrator obabo temp 113.15 thermostat PILE_L 1234 tau 1.0
+
+Example input files are provided in the examples/PACKAGES/pimd_bosonic directory.
+
+Description
+"""""""""""
+
+These fix commands are based on the fixes :doc:`pimd/nvt and
```
Comment (sjplimp) [inline doc/src/fix_pimd_bosonic.rst:67]: These 2 paragraphs are great context for users - thanks

File: doc/src/fix_pimd_bosonic.rst [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+Restrictions
+""""""""""""
+
+These fixes are part of the REPLICA package.  They are only enabled if
+LAMMPS was built with that package.  See the :doc:`Build package
+<Build_package>` page for more info.
+
+The restrictions of :doc:`fix pimd <fix_pimd>` apply.
+
+Default
```
Comment (sjplimp) [inline doc/src/fix_pimd_bosonic.rst:202]: Typically, all command doc pages have a "Related commands" section after Restrictions and before Default.
The one here should provide links to the 2 older PIMD commands.  And you should add links to your 2 new commands to the Related section on the old doc page.

**Comment by sjplimp** [comment]
Thanks for all the changes - the added doc page Description paragraphs read great.  Just a couple minor things left, where I added comments.

**Comment by OfirBlumer** [comment]
Thanks @sjplimp, I fixed the minor issue on fix.rst and added "related commands" sections on the two pimd doc files.

**Review by sjplimp — APPROVED** [review]

**Comment by yotamfe** [comment]
Hi @akohlmey,

@OfirBlumer and I noticed that the documentation for `fix pimd/langevin/bosonic` and `fix pimd/nvt/bosonic` is missing - the bosonic fixes mentioned on the [documentation](https://docs.lammps.org/fix_pimd.html), but their description on a separate page seems to have been removed in [this commit](https://github.com/lammps/lammps/commit/fcb2eee6863455e8b63b9f690cc7ee8920b60382). Is there anything we should do?

Thanks again for reviewing and accepting our contribution!
Yotam

**Comment by akohlmey** [comment]
> Is there anything we should do?

We try to avoid having separate documentation pages with such highly redundant content and rather merge them. If you feel that my merging is omitted something, please submit a pull request for the combined doc page.

**Comment by yotamfe** [comment]
Added in PR #4657. Thanks!
